Tighten parameter checks

cgmb opened the pull request:
When I first starting using node-argon2, I noticed a couple issues. It turned out that you fixed them (98d639ca14818d8514b39428f2f38f7aca65c6e8, a604e2a3c30a785e13309ffaa8590c5a22acad89) before I even got a chance to file an issue, but I decided nevertheless to review the node-argon2 source to see if I could spot any further problems.

The good news is, I didn't find anything particularly worrisome. Most of the problems I found were around edge cases that I'd probably never encounter. Nevertheless, they should be fixed.

This pull request tightens up the validation done on parameters.

1. The maximum memory cost is reported as 32, but supplying the value 32 is an error. I've reduced the reported maximum to 31.

2. Input values are not checked for overflow. This means that if you supply a ridiculously large cost for your hash, node-argon2 could possibly create a hash with a very small cost. I'm not sure if supplying such a ridiculously large cost actually makes any sense, but node-argon2 should always either do the right thing or return an error.

3. Negative input values are also accepted. If these negative values are large enough, they might be accepted without error as positive values. node-argon2 should probably just return an error when supplied with negative parameter values.

4. libargon2 could fail to hash for many reasons. For example, by providing a time cost that is too small. No error is reported by node-argon2 when this occurred. The hash would just be undefined. I've changed this to relay the error.

cgmb commented:
Actually, I'm just going to roll these in here, because 5 is related to the range checks and 6 is incredibly minor.

5) It's a rare platform where `unsigned int` is less than 32 bits, but since node-argon2 is getting its parameters as `uint32_t` and passing them to argon2 as `uint32_t`, we might as well just store them as `uint32_t`.

6) This is of little consequence,  but you seem to be trying to use the std versions of the C functions. In one case, you appear to be using the C version by accident. You get away with it because cstring is allowed to define both `std::strlen` and `strlen` [[1](http://stackoverflow.com/a/32606280/331041)].

ranisalt commented:
> Negative input values are also accepted. If these negative values are large enough, they might be accepted without error as positive values. node-argon2 should probably just return an error when supplied with negative parameter values.

I didn't remember to check it since it's only an issue on the Javascript side, as the C++ side only uses unsigned values, thanks!

> libargon2 could fail to hash for many reasons. For example, by providing a time cost that is too small. No error is reported by node-argon2 when this occurred. The hash would just be undefined. I've changed this to relay the error.

I wanted to get the error message directly from [here](https://github.com/P-H-C/phc-winner-argon2/blob/master/src/argon2.c#L23), check it it's possible to get the error message using [this function](https://github.com/P-H-C/phc-winner-argon2/blob/master/src/argon2.c#L389-L402) instead. I also prefer to have the verbatim message and not prepending with "Hashing failed with...".

cgmb commented:
Ok. Three changes.

> I wanted to get the error message directly from here, check it it's possible to get the error message using this function instead. I also prefer to have the verbatim message and not prepending with "Hashing failed with...".

I entirely agree. Somehow, I missed that when I skimmed over the argon2 interface. Fixed.

I tweaked a few test names, too. I noticed I'd been inconsistent with a couple of them.

I also dropped the `LCOV_EXCL_****` statements, as these lines are being tested.

ranisalt commented on test.spec.js:
```diff
@@ -118,6 +118,51 @@ module.exports = {
+
+    argon2.hash("password", "somesalt", {
+      timeCost: 4294967297
+    }, function (err, hash) {
```
You used 4294967296 every other place, could you please use the same value?

cgmb replied:
I chose 4294967297 because it becomes 1 when converted to uint32, while 4294967296 becomes 0. The value 1 is a valid input parameter for time cost, while 0 is not. Thus, if you remove the validation logic, the test fails with a more obviously concerning error when using 4294967297.

I was definitely inconsistent about it, though. Perhaps I should instead switch the other tests to ensure they overflow to valid inputs? Not a big deal either way. I could just switch them all to 4294967296 if you'd still prefer that.

cgmb commented:
Is it ok to set the return value after calling `Nan::ThrowTypeError`? That's done [here](https://github.com/ranisalt/node-argon2/blob/1ec49efc3ab49488deee22c2085a95a7cf64470a/src/argon2_node.cpp#L109), and I copied that style for my checks.

The function to throw the exception is generated by a macro in nan, and its definition [can be found here](https://github.com/nodejs/nan/blob/eca212b486d2dcf009201291472911dd842df623/nan.h#L649).

Following that into the v8 source I noticed [some interesting comments](https://github.com/v8/v8/blob/03ce7711e474a0ef74f723b30ae1527c89dec010/include/v8.h#L5664):
```
  /**
   * Schedules an exception to be thrown when returning to JavaScript.  When an
   * exception has been scheduled it is illegal to invoke any JavaScript
   * operation; the caller must return immediately and only after the exception
   * has been handled does it become legal to invoke JavaScript operations.
   */
  Local<Value> ThrowException(Local<Value> exception);
```

However, it's not entirely clear to me whether setting the return value like that is 'a JavaScript operation'. Unfortunately, I'm not too familiar with v8.

ranisalt commented:
Actually it makes no difference to set the return value, I just set it for the sake of meaning. Throwing an error means the return value won't even be checked anyway. I based it on node's implementations of bcrypt and they do it in this order, throw then set return type.

cgmb commented:
I fixed a couple problems in the synchronous tests:

1. I was inconsistent about the large number I used. I switched both time cost and parallelism to always use 4294967297.
2. The tests actually were throwing exceptions because I was still calling `encryptSync`, and were passing because an exception was expected. I now check the exception's error message to ensure it's the right error.

ranisalt commented:
Thanks! Please keep it up and report/PR if you find any improvements again :smile:
